tuist/tuist
# Support preferredScreenCaptureFormat in test options

### Note
- Based on the XcodeProj PR https://github.com/tuist/XcodeProj/pull/792
- Will require to update the XcodeProj dependencies in `Package.swift` and `Tuist/Dependencies.swift` when the new version of XcodeProj is released.

### Short description 📝

- Xcode 15 supports automatic screen recordings for UI tests.
- From [Xcode 15 beta Release Notes](https://developer.apple.com/documentation/xcode-release-notes/xcode-15-release-notes#New-Features-in-Xcode-15-Beta):
> XCTest now supports automatic screen recordings in addition to screenshots. Screen recordings allow for a fine-grained view of what happened during a test run, which can be meaningful to investigate UI test failures. Screen recordings are enabled by default (in favor of screenshots). They can be disabled in the test plan or the scheme’s Test action options.
- Available options are:
  - Screenshots - prior to Xcode 15 is was the only available format
  - Screen Recording - new feature of Xcode 15
- It can be configured in the scheme test action options
   <details><summary> New option screenshot </summary>
   <img width="928" alt="Screenshot 2023-09-01 at 17 21 35" src="https://github.com/tuist/XcodeProj/assets/5187973/e1848a46-e881-474c-a7b3-03f9a24e25ce">
   </details>

### How to test the changes locally 🧐

There are three paths:
- Explicit Test Action option
  Example:
  - Ensure `preferredScreenCaptureFormat` is set to `.screenshots` in line 38 of `projects/tuist/fixtures/ios_app_with_custom_scheme/Workspace.swift`
  - Run `swift run tuist generate --path projects/tuist/fixtures/ios_app_with_custom_scheme`
  - Open the generated workspace with Xcode 15
  - Open "Workspace-App" scheme editor
  - Ensure test options have `Automatically capture` set to `screenshots` 
  - Open the generated workspace with Xcode 14
  - Ensure test options are not affected, `Capture screenshots automatically` is selected
  - Repeat test with `preferredScreenCaptureFormat: .screenRecording`
- Project autogenerated scheme options
  Example:
  - Ensure `preferredScreenCaptureFormat`  is set to `.screenshots` in line 10 of `projects/tuist/fixtures/ios_app_with_tests/Project.swift`
  - Run `swift run tuist generate --path projects/tuist/fixtures/ios_app_with_tests `
  - Open the generated project with Xcode 15
  - Open "AppTests" scheme editor
  - Ensure test options have `Automatically capture` set to `screenshots` 
  - Check other scheme, containing tests: "AppUITests", "AppCoreTests", "MacFrameworkTests"
  - Open the generated workspace with Xcode 14
  - Ensure test options are not affected, `Capture screenshots automatically` is selected
  - Repeat test with `testScreenCaptureFormat: .screenRecording`
- Workspace autogenerated scheme options
  Example:
  - Add `, testScreenCaptureFormat: .screenshots` to line 9 of `projects/tuist/fixtures/workspace_with_test_locale/Workspace.swift`
  - Run `swift run tuist generate --path projects/tuist/fixtures/workspace_with_test_locale`
  - Open the generated workspace with Xcode 15
  - Open "App-Workspace" scheme editor
  - Ensure test options have `Automatically capture` set to `screenshots` 
  - Open the generated workspace with Xcode 14
  - Ensure test options are not affected, `Capture screenshots automatically` is selected
  - Repeat test with `testScreenCaptureFormat: .screenRecording`

### Contributor checklist ✅


### Reviewer checklist ✅
